src/metabase/lib/join.cljc

```diff
@@ -580,11 +592,67 @@
      (when-let [fk-col (fk-column-for query stage-number pk-col)]
        (lib.filter/filter-clause (equals-join-condition-operator-definition) fk-col pk-col)))))
 
-(mu/defn joined-thing :- [:maybe [:or lib.metadata/TableMetadata lib.metadata/CardMetadata]]
+(def ^:private Joinable
+  [:or lib.metadata/TableMetadata lib.metadata/CardMetadata])
+
+(mu/defn joined-thing :- [:maybe Joinable]
   "Return metadata about the origin of `a-join` using `metadata-providerable` as the source of information."
   [metadata-providerable :- lib.metadata/MetadataProviderable
    a-join                :- ::lib.schema.join/join]
   (let [origin (-> a-join :stages first)]
     (cond
       (:source-card origin)  (lib.metadata/card metadata-providerable (:source-card origin))
       (:source-table origin) (lib.metadata/table metadata-providerable (:source-table origin)))))
+
+(defn- add-join-alias-to-joinable-columns [cols a-join]
+  (let [join-alias     (current-join-alias a-join)
+        unique-name-fn (lib.util/unique-name-generator)]
+    (mapv (fn [col]
+            (as-> col col
+              (with-join-alias col join-alias)
+              (add-source-and-desired-aliases a-join unique-name-fn col)))
+          cols)))
+
+(defn- mark-selected-joinable-columns
+  "Mark the column metadatas in `cols` as `:selected` if they appear in `a-join`'s `:fields`."
+  [cols a-join]
+  (let [j-fields (join-fields a-join)]
+    (case j-fields
+      :all        (mapv (fn [col]
+                          (assoc col :selected? true))
+                        cols)
+      (:none nil) (mapv (fn [col]
+                          (assoc col :selected? false))
+                        cols)
```

Comment:
I know we normally only include `:selected` if its truthy, but I found that it was much easier to write my tests to ensure something was **not** selected if I could assert that the key was there and was false as opposed to having `=?` possibly match something that had an unexpected `:selected` key.